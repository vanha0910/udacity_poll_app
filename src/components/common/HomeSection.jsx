import React from "react";
import QuestionCard from "./QuestionCard";

const HomeSection = ({ data, title }) => {
  return (
    <div className="section">
      <div className="section-header">{title}</div>
      <div className="section-container">
        {data &&
          data.map((question) => (
            <QuestionCard key={question.id} question={question} />
          ))}
      </div>
    </div>
  );
};

export default HomeSection;
