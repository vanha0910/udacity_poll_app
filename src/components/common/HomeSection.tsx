import React from "react";
import QuestionCard from "./QuestionCard";
import { QuestionWithAuthor } from "../../redux/type";

interface HomeSectionProps {
  data: QuestionWithAuthor[];
  title: string;
}

const HomeSection = ({ data, title }: HomeSectionProps) => {
  return (
    <div className="section">
      <div className="section-header">{title}</div>
      <div className="section-container">
        {data &&
          data.map((question) => (
            <QuestionCard key={question.id} question={question} />
          ))}
      </div>
    </div>
  );
};

export default HomeSection;
