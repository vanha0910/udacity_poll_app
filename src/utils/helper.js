export const formatQuestionWithFullAuthorInfo = (question, users) => {
  return {
    ...question,
    author: users ? users.find((user) => user.id === question.author) : null,
  };
};

export const sortUserToTotalAnsweredAndQuestions = (users) => {
  return [...users].sort((a, b) => {
    const totalA = Object.keys(a.answers).length + a.questions.length;
    const totalB = Object.keys(b.answers).length + b.questions.length;
    return totalB - totalA;
  });
};

export const formatAnsweredQuestion = (question, user) => {
  const optionOneSelected =
    question.optionOne && question.optionOne.votes.includes(user);
  const optionTwoSelected =
    question.optionTwo && question.optionTwo.votes.includes(user);
  const isAnswered = !!(optionOneSelected || optionTwoSelected);
  let optionOne = {
    ...question.optionOne,
  };
  let optionTwo = {
    ...question.optionTwo,
  };
  if (isAnswered) {
    optionOne = {
      ...optionOne,
      count: question.optionOne.votes.length,
      percentage: (
        (100 * question.optionOne.votes.length) /
        (question.optionOne.votes.length + question.optionTwo.votes.length)
      ).toFixed(2),
    };
    optionTwo = {
      ...optionTwo,
      count: question.optionTwo.votes.length,
      percentage: (
        (100 * question.optionTwo.votes.length) /
        (question.optionOne.votes.length + question.optionTwo.votes.length)
      ).toFixed(2),
    };
  }
  return {
    ...question,
    optionOne,
    optionTwo,
    isAnswered,
    optionOneSelected,
    optionTwoSelected,
  };
};

export const formatTimestamp = (timestamp) => {
  const date = new Date(timestamp);

  const hours = date.getHours();
  const minutes = date.getMinutes();

  const formattedTime = hours >= 12 ? "PM" : "AM";
  const formattedHours = hours % 12 === 0 ? 12 : hours % 12;
  const formattedMinutes = minutes < 10 ? "0" + minutes : minutes;
  const timeString = `${formattedHours}:${formattedMinutes} ${formattedTime}`;

  const month = date.getMonth() + 1; // Months are 0-based
  const day = date.getDate();
  const year = date.getFullYear();

  const dateString = `${month}/${day}/${year}`;
  return `${timeString} | ${dateString}`;
};
