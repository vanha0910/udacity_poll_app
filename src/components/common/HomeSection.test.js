import { renderWithProviders } from "../../utils/testUtils";
import React from "react";
import { MemoryRouter } from "react-router-dom";
import HomeSection from "./HomeSection";

describe("HomeSection Component", () => {
  it("matches snapshot", () => {
    const { container } = renderWithProviders(
      <MemoryRouter>
        <HomeSection />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });
});
