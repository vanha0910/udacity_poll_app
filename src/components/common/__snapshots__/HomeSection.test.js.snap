// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeSection Component matches snapshot 1`] = `
<div>
  <div
    class="section"
  >
    <div
      class="section-header"
    />
    <div
      class="section-container"
    />
  </div>
</div>
`;

exports[`renders Header component correctly 1`] = `
<div>
  <header
    class="header"
  >
    <nav
      class="nav"
    >
      <ul
        class="nav-list"
      >
        <li
          class="nav-item  active"
        >
          <a
            class="nav-link"
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="nav-item "
        >
          <a
            class="nav-link"
            href="/leaderboard"
          >
            Leaderboard
          </a>
        </li>
        <li
          class="nav-item "
        >
          <a
            class="nav-link"
            href="/add"
          >
            New
          </a>
        </li>
      </ul>
    </nav>
    <div
      class="user-info"
    >
      <button
        class="logout-btn"
      >
        Logout
      </button>
    </div>
  </header>
</div>
`;

exports[`renders HomeSection component correctly 1`] = `
<div>
  <div
    class="section"
  >
    <div
      class="section-header"
    />
    <div
      class="section-container"
    />
  </div>
</div>
`;
